### ISSUE TITLE
UtBS S1: bad_function_call exception thrown in turn_end event

### ISSUE THREAD

**Issue by soliton-** [wesnoth/wesnoth#9237]
### Game and System Information

- **Version:** 1.19.3-20-g9b70484b4f4
- **OS:** macOS

### Description of the bug

``` * thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x000000019d7ad5f0 libsystem_kernel.dylib`__pthread_kill + 8
    frame #1: 0x000000010fc7abe8 libsystem_pthread.dylib`pthread_kill + 288
    frame #2: 0x000000019d6f2a30 libsystem_c.dylib`abort + 180
    frame #3: 0x000000019d79cd08 libc++abi.dylib`abort_message + 132
    frame #4: 0x000000019d78cfa4 libc++abi.dylib`demangling_terminate_handler() + 320
    frame #5: 0x000000019d427c00 libobjc.A.dylib`_objc_terminate() + 160
    frame #6: 0x000000019d79c0cc libc++abi.dylib`std::__terminate(void (*)()) + 16
    frame #7: 0x000000019d79c070 libc++abi.dylib`std::terminate() + 108
    frame #8: 0x0000000102c79514 The Battle for Wesnoth`__clang_call_terminate + 12
    frame #9: 0x00000001038328f4 The Battle for Wesnoth`std::__1::unique_ptr<lua_formula_bridge::fwrapper, std::__1::function<void (lua_formula_bridge::fwrapper*)>>::reset[abi:ue170006](this=0x000000016d17ff80, __p=0x0000000000000000) at unique_ptr.h:300:7
    frame #10: 0x0000000103832c14 The Battle for Wesnoth`std::__1::unique_ptr<lua_formula_bridge::fwrapper, std::__1::function<void (lua_formula_bridge::fwrapper*)>>::~unique_ptr[abi:ue170006](this=0x000000016d17ff80) at unique_ptr.h:266:75
    frame #11: 0x0000000103832168 The Battle for Wesnoth`std::__1::unique_ptr<lua_formula_bridge::fwrapper, std::__1::function<void (lua_formula_bridge::fwrapper*)>>::~unique_ptr[abi:ue170006](this=0x000000016d17ff80) at unique_ptr.h:266:73
    frame #12: 0x00000001044bc818 The Battle for Wesnoth`lua_formula_bridge::intf_eval_formula(L=0x0000000143a13008) at lua_formula_bridge.cpp:268:1
    frame #13: 0x000000010468ddec The Battle for Wesnoth`precallC(L=0x0000000143a13008, func=0x0000000143cfbdd0, nresults=1, f=(The Battle for Wesnoth`lua_formula_bridge::intf_eval_formula(lua_State*) at lua_formula_bridge.cpp:249)) at ldo.c:529:7
    frame #14: 0x000000010468deb0 The Battle for Wesnoth`luaD_precall(L=0x0000000143a13008, func=0x0000000143cfbdd0, nresults=1) at ldo.c:595:7
    frame #15: 0x00000001046bab04 The Battle for Wesnoth`luaV_execute(L=0x0000000143a13008, ci=0x0000600003c5c500) at lvm.c:1684:22
    frame #16: 0x000000010468e24c The Battle for Wesnoth`ccall(L=0x0000000143a13008, func=0x0000000143cfbda0, nResults=1, inc=65537) at ldo.c:637:5
    frame #17: 0x000000010468e29c The Battle for Wesnoth`luaD_callnoyield(L=0x0000000143a13008, func=0x0000000143cfbda0, nResults=1) at ldo.c:655:3
    frame #18: 0x000000010467cc24 The Battle for Wesnoth`f_call(L=0x0000000143a13008, ud=0x000000016d181108) at lapi.c:1038:3
    frame #19: 0x000000010468c5f0 The Battle for Wesnoth`luaD_rawrunprotected(L=0x0000000143a13008, f=(The Battle for Wesnoth`f_call(lua_State*, void*) at lapi.c:1036), ud=0x000000016d181108) at ldo.c:144:3
    frame #20: 0x000000010468ea1c The Battle for Wesnoth`luaD_pcall(L=0x0000000143a13008, func=(The Battle for Wesnoth`f_call(lua_State*, void*) at lapi.c:1036), u=0x000000016d181108, old_top=1440, ef=1424) at ldo.c:953:12
    frame #21: 0x000000010467cac8 The Battle for Wesnoth`lua_pcallk(L=0x0000000143a13008, nargs=1, nresults=1, errfunc=-3, ctx=0, k=0x0000000000000000) at lapi.c:1064:14
    frame #22: 0x0000000103a591c4 The Battle for Wesnoth`luaW_pcall_internal(L=0x0000000143a13008, nArgs=1, nRets=1) at lua_common.cpp:1119:16
    frame #23: 0x0000000103a59248 The Battle for Wesnoth`luaW_pcall(L=0x0000000143a13008, nArgs=1, nRets=1, allow_wml_error=true) at lua_common.cpp:1136:12
  * frame #24: 0x00000001037ab09c The Battle for Wesnoth`game_lua_kernel::run_wml_conditional(this=0x0000000124331930, cmd="variable", cfg=0x000000016d181b18) at game_lua_kernel.cpp:5743:6
    frame #25: 0x0000000103172c2c The Battle for Wesnoth`game_events::(anonymous namespace)::internal_conditional_passed(cond=0x000000016d181eb8) at conditional_wml.cpp:179:32
    frame #26: 0x000000010317249c The Battle for Wesnoth`game_events::conditional_passed(cond=0x000000016d181eb8) at conditional_wml.cpp:193:17
    frame #27: 0x00000001037a3bcc The Battle for Wesnoth`intf_eval_conditional(L=0x0000000143a13008) at game_lua_kernel.cpp:1830:11
    frame #28: 0x000000010468ddec The Battle for Wesnoth`precallC(L=0x0000000143a13008, func=0x0000000143cfbd70, nresults=1, f=(The Battle for Wesnoth`intf_eval_conditional(lua_State*) at game_lua_kernel.cpp:1828)) at ldo.c:529:7
    frame #29: 0x000000010468deb0 The Battle for Wesnoth`luaD_precall(L=0x0000000143a13008, func=0x0000000143cfbd70, nresults=1) at ldo.c:595:7
    frame #30: 0x00000001046bab04 The Battle for Wesnoth`luaV_execute(L=0x0000000143a13008, ci=0x0000600003db4740) at lvm.c:1684:22
    frame #31: 0x000000010468e24c The Battle for Wesnoth`ccall(L=0x0000000143a13008, func=0x0000000143cfb850, nResults=0, inc=65537) at ldo.c:637:5
    frame #32: 0x000000010468e29c The Battle for Wesnoth`luaD_callnoyield(L=0x0000000143a13008, func=0x0000000143cfb850, nResults=0) at ldo.c:655:3
    frame #33: 0x000000010467c980 The Battle for Wesnoth`lua_callk(L=0x0000000143a13008, nargs=1, nresults=0, ctx=0, k=0x0000000000000000) at lapi.c:1020:5
    frame #34: 0x00000001037ab29c The Battle for Wesnoth`intf_run_event_wml(L=0x0000000143a13008) at game_lua_kernel.cpp:5760:2
    frame #35: 0x000000010468ddec The Battle for Wesnoth`precallC(L=0x0000000143a13008, func=0x0000000143cfb830, nresults=0, f=(The Battle for Wesnoth`intf_run_event_wml(lua_State*) at game_lua_kernel.cpp:5754)) at ldo.c:529:7
    frame #36: 0x000000010468deb0 The Battle for Wesnoth`luaD_precall(L=0x0000000143a13008, func=0x0000000143cfb830, nresults=0) at ldo.c:595:7
    frame #37: 0x000000010468e21c The Battle for Wesnoth`ccall(L=0x0000000143a13008, func=0x0000000143cfb830, nResults=0, inc=65537) at ldo.c:635:13
    frame #38: 0x000000010468e29c The Battle for Wesnoth`luaD_callnoyield(L=0x0000000143a13008, func=0x0000000143cfb830, nResults=0) at ldo.c:655:3
    frame #39: 0x000000010467cc24 The Battle for Wesnoth`f_call(L=0x0000000143a13008, ud=0x000000016d1831b8) at lapi.c:1038:3
    frame #40: 0x000000010468c5f0 The Battle for Wesnoth`luaD_rawrunprotected(L=0x0000000143a13008, f=(The Battle for Wesnoth`f_call(lua_State*, void*) at lapi.c:1036), ud=0x000000016d1831b8) at ldo.c:144:3
    frame #41: 0x000000010468ea1c The Battle for Wesnoth`luaD_pcall(L=0x0000000143a13008, func=(The Battle for Wesnoth`f_call(lua_State*, void*) at lapi.c:1036), u=0x000000016d1831b8, old_top=48, ef=32) at ldo.c:953:12
    frame #42: 0x000000010467cac8 The Battle for Wesnoth`lua_pcallk(L=0x0000000143a13008, nargs=1, nresults=0, errfunc=-3, ctx=0, k=0x0000000000000000) at lapi.c:1064:14
    frame #43: 0x0000000103a591c4 The Battle for Wesnoth`luaW_pcall_internal(L=0x0000000143a13008, nArgs=1, nRets=0) at lua_common.cpp:1119:16
    frame #44: 0x0000000103a59248 The Battle for Wesnoth`luaW_pcall(L=0x0000000143a13008, nArgs=1, nRets=0, allow_wml_error=true) at lua_common.cpp:1136:12
    frame #45: 0x00000001037ab8e0 The Battle for Wesnoth`game_lua_kernel::run_wml_event(this=0x0000000124331930, ref=63, args=0x000000016d1839f0, ev=0x0000600001038580, out=0x0000000000000000) at game_lua_kernel.cpp:5833:5
    frame #46: 0x000000010391c31c The Battle for Wesnoth`game_events::event_handler::handle_event(this=0x0000600001279a58, event_info=0x0000600001038580, lk=0x0000000124331930) at handlers.cpp:115:5
    frame #47: 0x0000000103d92f88 The Battle for Wesnoth`game_events::wml_event_pump::process_event(this=0x00006000025effe0, handler_p=std::__1::shared_ptr<game_events::event_handler>::element_type @ 0x0000600001279a58 strong=3 weak=2, ev=0x0000600001038580) at pump.cpp:218:13
    frame #48: 0x0000000103d9fffc The Battle for Wesnoth`game_events::wml_event_pump::operator()()::$_0::operator()(this=0x000060000238ff28, (null)=0x0000600003da8f00, ptr=std::__1::shared_ptr<game_events::event_handler>::element_type @ 0x0000600001279a58 strong=3 weak=2) const at pump.cpp:498:5
    frame #49: 0x0000000103d9fe88 The Battle for Wesnoth`decltype(std::declval<game_events::wml_event_pump::operator()()::$_0&>()(std::declval<game_events::manager&>(), std::declval<std::__1::shared_ptr<game_events::event_handler>&>())) std::__1::__invoke[abi:ue170006]<game_events::wml_event_pump::operator()()::$_0&, game_events::manager&, std::__1::shared_ptr<game_events::event_handler>&>(__f=0x000060000238ff28, __args=0x0000600003da8f00, __args=std::__1::shared_ptr<game_events::event_handler>::element_type @ 0x0000600001279a58 strong=3 weak=2) at invoke.h:340:25
    frame #50: 0x0000000103d9fe30 The Battle for Wesnoth`void std::__1::__invoke_void_return_wrapper<void, true>::__call[abi:ue170006]<game_events::wml_event_pump::operator()()::$_0&, game_events::manager&, std::__1::shared_ptr<game_events::event_handler>&>(__args=0x000060000238ff28, __args=0x0000600003da8f00, __args=std::__1::shared_ptr<game_events::event_handler>::element_type @ 0x0000600001279a58 strong=3 weak=2) at invoke.h:415:5
    frame #51: 0x0000000103d9fdfc The Battle for Wesnoth`std::__1::__function::__alloc_func<game_events::wml_event_pump::operator()()::$_0, std::__1::allocator<game_events::wml_event_pump::operator()()::$_0>, void (game_events::manager&, std::__1::shared_ptr<game_events::event_handler>&)>::operator()[abi:ue170006](this=0x000060000238ff28, __arg=0x0000600003da8f00, __arg=std::__1::shared_ptr<game_events::event_handler>::element_type @ 0x0000600001279a58 strong=3 weak=2) at function.h:193:16
    frame #52: 0x0000000103d9ede8 The Battle for Wesnoth`std::__1::__function::__func<game_events::wml_event_pump::operator()()::$_0, std::__1::allocator<game_events::wml_event_pump::operator()()::$_0>, void (game_events::manager&, std::__1::shared_ptr<game_events::event_handler>&)>::operator()(this=0x000060000238ff20, __arg=0x0000600003da8f00, __arg=std::__1::shared_ptr<game_events::event_handler>::element_type @ 0x0000600001279a58 strong=3 weak=2) at function.h:364:12
    frame #53: 0x0000000103aeb3c8 The Battle for Wesnoth`std::__1::__function::__value_func<void (game_events::manager&, std::__1::shared_ptr<game_events::event_handler>&)>::operator()[abi:ue170006](this=0x000000016d184d08, __args=0x0000600003da8f00, __args=std::__1::shared_ptr<game_events::event_handler>::element_type @ 0x0000600001279a58 strong=3 weak=2) const at function.h:518:16
    frame #54: 0x0000000103ae7ce8 The Battle for Wesnoth`std::__1::function<void (game_events::manager&, std::__1::shared_ptr<game_events::event_handler>&)>::operator()(this= Lambda in File pump.cpp at Line 494, __arg=0x0000600003da8f00, __arg=std::__1::shared_ptr<game_events::event_handler>::element_type @ 0x0000600001279a58 strong=3 weak=2) const at function.h:1169:12
    frame #55: 0x0000000103ae7a74 The Battle for Wesnoth`game_events::manager::execute_on_events(this=0x0000600003da8f00, event_id="turn_end", func= Lambda in File pump.cpp at Line 494) at manager.cpp:234:6
    frame #56: 0x0000000103d94c84 The Battle for Wesnoth`game_events::wml_event_pump::operator()(this=0x00006000025effe0) at pump.cpp:494:23
    frame #57: 0x0000000103d9429c The Battle for Wesnoth`game_events::wml_event_pump::fire(this=0x00006000025effe0, event="turn_end", loc1=0x0000000104c47860, loc2=0x0000000104c47860, data=0x000000016d184e70) at pump.cpp:403:9
    frame #58: 0x0000000103d4f8e4 The Battle for Wesnoth`play_controller::finish_turn(this=0x000000016d185908) at play_controller.cpp:595:9
    frame #59: 0x0000000103d81e30 The Battle for Wesnoth`playsingle_controller::finish_side_turn(this=0x000000016d185908) at playsingle_controller.cpp:279:3
    frame #60: 0x0000000103d81a58 The Battle for Wesnoth`playsingle_controller::play_some(this=0x000000016d185908) at playsingle_controller.cpp:218:3
    frame #61: 0x0000000103d82e00 The Battle for Wesnoth`playsingle_controller::play_scenario_main_loop(this=0x000000016d185908) at playsingle_controller.cpp:316:4
    frame #62: 0x0000000103d83788 The Battle for Wesnoth`playsingle_controller::play_scenario(this=0x000000016d185908, level=0x00000001438c63e8) at playsingle_controller.cpp:447:3
    frame #63: 0x0000000103d718b8 The Battle for Wesnoth`campaign_controller::playsingle_scenario(this=0x000000016d186580, end_level=0x000000016d1864c0) at playcampaign.cpp:66:42
    frame #64: 0x0000000103d71cd0 The Battle for Wesnoth`campaign_controller::play_game(this=0x000000016d186580) at playcampaign.cpp:136:11
    frame #65: 0x0000000103768f0c The Battle for Wesnoth`game_launcher::launch_game(this=0x00000001438c6000, reload=RELOAD_DATA) at game_launcher.cpp:1005:15
    frame #66: 0x0000000104435290 The Battle for Wesnoth`do_gameloop(cmdline_opts=0x000000016d189170) at wesnoth.cpp:892:10
    frame #67: 0x0000000104430c98 The Battle for Wesnoth`wesnoth_main(argc=2, argv=0x000060000276a720) at wesnoth.cpp:966:19
    frame #68: 0x0000000103f6fb88 The Battle for Wesnoth`-[SDLMain applicationDidFinishLaunching:](self=0x000060000255ee40, _cmd="applicationDidFinishLaunching:", note="NSApplicationDidFinishLaunchingNotification") at SDLMain.mm:106:11
    frame #69: 0x000000019d8ba144 CoreFoundation`__CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 148
    frame #70: 0x000000019d94e3d8 CoreFoundation`___CFXRegistrationPost_block_invoke + 88
    frame #71: 0x000000019d94e320 CoreFoundation`_CFXRegistrationPost + 440
    frame #72: 0x000000019d888678 CoreFoundation`_CFXNotificationPost + 768
    frame #73: 0x000000019e9a52c4 Foundation`-[NSNotificationCenter postNotificationName:object:userInfo:] + 88
    frame #74: 0x00000001a112b370 AppKit`-[NSApplication _postDidFinishNotification] + 284
    frame #75: 0x00000001a112b120 AppKit`-[NSApplication _sendFinishLaunchingNotification] + 172
    frame #76: 0x00000001a1129674 AppKit`-[NSApplication(NSAppleEventHandling) _handleAEOpenEvent:] + 504
    frame #77: 0x00000001a1129270 AppKit`-[NSApplication(NSAppleEventHandling) _handleCoreEvent:withReplyEvent:] + 492
    frame #78: 0x000000019e9ce154 Foundation`-[NSAppleEventManager dispatchRawAppleEvent:withRawReply:handlerRefCon:] + 316
    frame #79: 0x000000019e9cdf48 Foundation`_NSAppleEventManagerGenericHandler + 80
    frame #80: 0x00000001a48e79c4 AE`___lldb_unnamed_symbol867 + 1624
    frame #81: 0x00000001a48e72ec AE`___lldb_unnamed_symbol866 + 44
    frame #82: 0x00000001a48e08a8 AE`aeProcessAppleEvent + 488
    frame #83: 0x00000001a807d02c HIToolbox`AEProcessAppleEvent + 68
    frame #84: 0x00000001a1122fd4 AppKit`_DPSNextEvent + 1440
    frame #85: 0x00000001a19194d0 AppKit`-[NSApplication(NSEventRouting) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 700
    frame #86: 0x00000001a1115ffc AppKit`-[NSApplication run] + 476
    frame #87: 0x0000000103f6fd0c The Battle for Wesnoth`main(argc=4, argv=0x000000016d18b298) at SDLMain.mm:148:2
    frame #88: 0x000000019d45b154 dyld`start + 2476

(lldb) print cfg.cfg_.debug()
(std::string) "as_type = weapon\nformula = index_of('once_per_turn', value.specials)!=-1\nname = possibly_opt_attack\n"
```

### Steps to reproduce the behavior

1. Start UtBS
2. Move south to trigger enemies
3. End turn

### Additional context

The trigger appears to be the turn end event here:
https://github.com/wesnoth/wesnoth/blob/9b70484b4f4a1b0dc9c87a625c0ab664bf98a4db/data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg#L394-L438

**Comment by Wedge009** [comment]
I got a crash-to-desktop with the message:
```
terminate called after throwing an instance of 'std::bad_function_call'
  what():  bad_function_call
Aborted
```

**Comment by Wedge009** [comment]
This appears to have been introduced with commit 5ed7f63edeb9c942df3736211db11511d68f5af8 by @CelticMinstrel, included from 1.19.2 onwards.
